File: TableParts/TableParts/Code/TableSource.cs
using System;
using UIKit;
using System.Collections.Generic;
using Foundation;

namespace Example_TableParts
{
	/// <summary>
	/// Combined DataSource and Delegate for our UITableView
	/// </summary>
	public class TableSource : UITableViewSource
	{
		//---- declare vars
		protected List<TableItemGroup> tableItems;
		protected string cellIdentifier = "TableCell";

		public TableSource (List<TableItemGroup> items)
		{
			tableItems = items;
		}
		
		#region data binding/display methods
		
		/// <summary>
		/// Called by the TableView to determine how many sections(groups) there are.
		/// </summary>
		public override nint NumberOfSections (UITableView tableView)
		{
			return tableItems.Count;
		}

		/// <summary>
		/// Called by the TableView to determine how many cells to create for that particular section.
		/// </summary>
		public override nint RowsInSection (UITableView tableview, nint section)
		{
			return tableItems[(int)section].Items.Count;
		}

		/// <summary>
		/// Called by the TableView to retrieve the header text for the particular section(group)
		/// </summary>
		public override string TitleForHeader (UITableView tableView, nint section)
		{
			return tableItems[(int)section].Name;
		}

		/// <summary>
		/// Called by the TableView to retrieve the footer text for the particular section(group)
		/// </summary>
		public override string TitleForFooter (UITableView tableView, nint section)
		{
			return tableItems[(int)section].Footer;
		}
		
		#endregion	
				
		#region user interaction methods
		
		public override void RowSelected (UITableView tableView, NSIndexPath indexPath)
		{
			new UIAlertView ("Row Selected"
				, tableItems[(int)indexPath.Section].Items[(int)indexPath.Row], null, "OK", null).Show ();
		}
		
		public override void RowDeselected (UITableView tableView, NSIndexPath indexPath)
		{
			Console.WriteLine ("Row " + indexPath.Row.ToString () + " deselected");	
		}
		
		#endregion	

		/// <summary>
		/// Called by the TableView to get the actual UITableViewCell to render for the particular section and row
		/// </summary>
		public override UITableViewCell GetCell (UITableView tableView, Foundation.NSIndexPath indexPath)
		{
			Console.WriteLine ("Calling Get Cell, isEditing:" + tableView.Editing);
			
			//---- declare vars
			UITableViewCell cell = tableView.DequeueReusableCell (cellIdentifier);
			
			//---- if there are no cells to reuse, create a new one
			if (cell == null)
				cell = new UITableViewCell (UITableViewCellStyle.Default, cellIdentifier);
			
			//---- set the item text
			cell.TextLabel.Text = tableItems[(int)indexPath.Section].Items[(int)indexPath.Row];
			
			return cell;
		}
	}
}

File: TableParts/TablePartsFS/TableSource.fs
namespace TablePartsFS
open System
open System.Linq
open UIKit
open System.Collections.Generic
open Foundation

open Conversion

type TableItemGroup = {Name : string; Footer : string; Items : string list}

type TableSource(tableItems: TableItemGroup array) =
   inherit UITableViewSource()

   let cellIdentifier = "TableCell"

   /// Called by the TableView to determine how many sections(groups) there are.
   override x.NumberOfSections (tableView) =
        implicit tableItems.Length

   /// Called by the TableView to determine how many cells to create for that particular section.
   override x.RowsInSection (tableview, section) =
       nint (tableItems.[int section].Items.Count())
      

   /// Called by the TableView to retrieve the header text for the particular section(group)
   override x.TitleForHeader (tableView, section) =
       let title = tableItems.[int section].Name
       title

   /// Called by the TableView to retrieve the footer text for the particular section(group)
   override x.TitleForFooter (tableView, section) =
       tableItems.[int section].Footer  

   override x.RowSelected (tableView, indexPath) =
       (new UIAlertView ("Row Selected", tableItems.[int indexPath.Section].Items.[int indexPath.Row], null, "OK", null)).Show ()
    
   override x.RowDeselected (tableView, indexPath) =
       printfn "Row %s deselected" <| indexPath.Row.ToString()

   /// Called by the TableView to get the actual UITableViewCell to render for the particular section and row
   override x.GetCell (tableView, indexPath) =
       printfn "Calling Get Cell, isEditing:%b" tableView.Editing
        
       // declare vars
       let cell =
           match tableView.DequeueReusableCell (cellIdentifier) with
           // if there are no cells to reuse, create a new one
           | null -> new UITableViewCell (UITableViewCellStyle.Default, cellIdentifier)
           | a -> a
       
       // set the item text
       cell.TextLabel.Text <- tableItems.[int indexPath.Section].Items.[int indexPath.Row]
       cell